### PR TITLE
feat(agent-mcp): broaden send_to_user to CSV/JSON/Markdown with type-aware previews

### LINKED ISSUE #461: send_to_user MCP: broaden supported file types and add graceful fallback for unsupported types
## Summary

The `mcp__claudette__send_to_user` tool currently only accepts a narrow allow-list of MIME types (PNG/JPEG/GIF/WebP/SVG, PDF, and `text/plain`), but the tool description and system-prompt nudge advertise it broadly enough ("send / share / show / give them a file… or text artifact") that the model reaches for it on common file types it can't actually deliver — e.g. CSVs.

A user reported that they asked the agent for a ~600-row CSV; the model picked `mcp__claudette__send_to_user`, the call did not render anything inline, and the user ended up asking the agent to write the file to disk so they could grab it manually. The model's choice of tool was reasonable given the description; the gap is on Claudette's side.

## Reproduction

1. Ask the agent to send / share you a file that isn't an image, PDF, or `.txt` (e.g. CSV, JSON, Markdown, log file).
2. Observe: the agent invokes `mcp__claudette__send_to_user`, but nothing renders inline. The bridge rejects the call (or the policy in `src/agent_mcp/tools/send_to_user.rs::policy` returns an error) and the user is left without the file.

## Where this lives in the code

- **Allow-list & size caps:** `src/agent_mcp/tools/send_to_user.rs` (`ALLOWED_IMAGE_TYPES`, `ALLOWED_DOCUMENT_TYPES`, `ALLOWED_TEXT_TYPES`, `MAX_*_BYTES`, and `policy()`).
- **Tool schema advertised to the model:** `src/agent_mcp/server.rs::tools_list_result` — `media_type` is a JSON-Schema `enum` over the same allow-list, plus a description that says "screenshot, generated image, PDF, small text file".
- **System-prompt nudge:** `src/agent_mcp/mod.rs::SYSTEM_PROMPT_NUDGE` — currently broad ("file, image, screenshot, PDF, or text artifact"), which encourages over-selection.
- **Inbound (compose-side) symmetry:** `src/ui/src/utils/attachmentValidation.ts` (`SUPPORTED_*_TYPES`, `MAX_*_SIZE`). The Rust policy doc comment explicitly states symmetry with inbound is the default — widening outbound likely means widening inbound too.

## Proposed direction (open to discussion)

Two complementary tracks:

**1. Broader supported file types.** Treat \"data/text artifacts\" as a first-class category alongside images and PDFs. Candidates:
- `text/csv`, `application/json`, `text/markdown`, `text/x-log`, `application/x-yaml`, generic `text/*` with a size cap.
- Each likely needs its own preview/render UX (see below). Size caps should be type-aware (a 5 MB CSV is a different beast from a 5 MB PNG).

**2. Graceful fallback for still-unsupported types.** When the model picks `send_to_user` for a type we genuinely don't want to inline (binary blob, oversized file, executable), the tool should:
- Return a clear, structured error to the model so it can fall back (e.g. \"this type isn't supported inline; write to disk and tell the user the path\").
- The error message in `policy()` already lists supported types — make sure that text propagates to the model unchanged so it can self-correct rather than silently failing.
- Consider a host-side fallback: drop the file in a known location (e.g. an attachments dir under the workspace) and surface a downloadable link in chat instead of inline rendering.

**3. Tighten the model's mental model.**
- Narrow `SYSTEM_PROMPT_NUDGE` and the tool `description` to match the actual allow-list (or widen the allow-list to match — pick one and align).
- The `inputSchema.media_type` enum currently constrains the model at decode time, which is good. Re-check after broadening that the enum stays exhaustive.

## UX considerations for non-image artifacts

- **CSV:** table preview (sortable header, virtualized rows), with a \"download\" affordance. Same component could power a future inbound-CSV-attachment story.
- **Markdown:** rendered preview using the existing markdown renderer; collapse if long.
- **Long text / logs:** inline expandable card with monospace font and copy/download buttons; don't expand inline if over N lines.
- **JSON:** pretty-printed with syntax highlighting and collapse; \"copy as JSON\" affordance.

The right answer is probably \"images stay inline, large textual artifacts get a compact card with a preview-on-click\" rather than trying to make everything render the same way.

## Open questions

- Do we widen first and add UX after, or land the better fallback (\"write to disk + announce\") first as a safety net while we design the table/markdown viewers?
- Should `send_to_user` be **disabled by default** until the broader set is supported and polished? It's currently enabled by default per `src/agent_mcp/mod.rs::is_builtin_plugin_enabled` (absent setting = enabled).
- Should the inbound (`attachmentValidation.ts`) rules widen in lockstep, preserving the symmetry the policy module's doc comment calls out?

## Related

- Original Discord conversation around CSV delivery (2026-04-27): model selected the tool correctly, but the file type wasn't supported, so the user fell back to \"write to /tmp and grab it manually.\"
- Built-in plugin registration: `src/agent_mcp/mod.rs::BUILTIN_PLUGINS` and the toggle key `builtin_plugin:send_to_user:enabled`.

### PR DESCRIPTION
**Pull request by jamesbrink**
Closes #461.

## Summary

`mcp__claudette__send_to_user` advertised broadly ("send / share / show / give them a file… or text artifact") but only accepted images, PDF, and `text/plain` — so the model picked it for CSVs and similar artifacts that the host then silently dropped.

This PR widens the supported set, makes the rejection path actionable, and adds type-aware preview cards.

### Outbound (agent → user)
- `ALLOWED_TEXT_TYPES` gains `text/csv`, `text/markdown`, `application/json` with per-type caps (CSV 2 MiB, others 1 MiB; `text/plain` bumped from 500 KiB → 1 MiB).
- `policy()` rejection text now includes the filename and a clear fallback instruction ("the file at … is on disk — tell the user that path") so the model self-corrects to "announce the path" for unsupported / oversized types instead of retrying.
- Tool `description` and `SYSTEM_PROMPT_NUDGE` align to the actual allow-list and explicitly direct the agent to announce-the-path for non-supported types.
- The schema `enum` regenerates from the slice contents — no separate schema change needed.

### Inbound (user → agent)
Preserves the policy module's documented symmetry invariant:
- `attachmentValidation.ts` widens the MIME allow-list (now includes SVG too) with per-type caps.
- `read_file_as_base64` maps `.csv` / `.md` / `.json` / `.svg` to specific MIME types rather than collapsing them to `text/plain`.
- `send_chat_message`'s allow-list, size table, and UTF-8 decoder follow the same shape.

### Inline preview UX
A new `MessageAttachment` switch routes attachment rendering by media type:
- **CSV** — parsed quote-aware, rendered as an HTML `<table>` with header row + first ~50 rows; "+ N more rows" footer when truncated.
- **Markdown** — rendered through `MessageMarkdown` so the chat surface's heading/list/code-block styling applies; collapsible past ~320 px.
- **JSON** — pretty-printed (2-space indent), with a "(invalid JSON — showing raw)" fallback.
- **Plain text** — monospace block in a collapsible card (replaces the previous 1-line filename+size badge so logs and other text artifacts get an actual preview).

Each card lazy-loads attachment bytes via a shared `useAttachmentText` hook (mirrors `PdfThumbnail`'s lazy strategy).

### Tests
- `policy_table` covers each new accepted type and oversize rejection; new `rejection_text_is_actionable` asserts the rejection mentions both the filename and the on-disk fallback.
- `attachments.test.ts` widened to cover the new types and caps; new `csvParse.test.ts` and `JsonAttachmentCard.test.ts` exercise the parsing logic.
- `cargo test`, `cargo clippy --workspace --all-targets`, `cargo fmt --check`, `bunx tsc -b`, `bun run test` all green.

### Open questions answered
- **Widen first or fallback first?** Widen — the structured rejection already reaches the model; we just made it actionable in the same change. No separate "drop-into-attachments-dir" host fallback is needed because the file is already on disk.
- **Default-on / default-off?** Default-on stays. Disabling would punish the 90% case for the 10% edge case; the existing `Plugins → Agent Attachments` toggle remains the opt-out.
- **Inbound symmetry?** Widened in lockstep — preserves the doc-comment invariant.

## Test plan

- [x] `cargo test --all-features` (741 passed)
- [x] `cd src/ui && bunx tsc -b`
- [x] `cd src/ui && bun run test` (948 passed)
- [x] `cargo clippy --workspace --all-targets` (zero warnings)
- [x] `cargo fmt --all --check`
- [ ] Manual UAT in dev build: ask the agent to send a CSV, JSON, Markdown, and plain-text file — verify each renders with the right card. (Skipped here because the running dev instance was wired to a different checkout; flagging for reviewer.)
- [ ] Manual UAT: ask the agent to send an unsupported type (e.g. `.tar.gz`); verify the new actionable error reaches the model and it announces the path.
- [ ] Manual UAT: drop a `.csv` / `.json` / `.md` into the composer and confirm inbound delivery still works.

### CODE REVIEW

File: src/ui/src/components/chat/AttachmentCardShell.tsx [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+};
+
+/**
+ * Common shell used by the CSV / Markdown / JSON / plain-text cards:
+ * a header (icon + filename + size) and a body that collapses past
+ * `collapsedHeightPx` with an "Expand" / "Collapse" toggle.
+ *
+ * `onContextMenu` lets the call site wire the same Download / Copy / Open
+ * menu image attachments use.
+ */
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/components/chat/AttachmentCardShell.tsx:37]: Addressed in a52bc70 — the doc comment now describes what the shell actually does: when `collapsible` is true, the body clamps to ~320px via the `.body { max-height: 320px }` rule in the CSS module, with an Expand/Collapse toggle. The phantom `collapsedHeightPx` prop reference is gone.

File: src/ui/src/utils/csvParse.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        i++;
+      }
+      // Don't push completely-blank lines that come from trailing newlines.
+      if (!(row.length === 1 && row[0] === "")) {
+        rows.push(row);
+        if (rows.length >= maxRows) {
+          return rows;
+        }
+      }
+      row = [];
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/utils/csvParse.ts]: Addressed in a52bc70 — `parseCsv` and the rename of `countCsvLines` → `countCsvRows` now apply the same blank-line rule everywhere. Both functions skip a row whose physical line has no commas and no content (regardless of position) and treat newlines inside quoted fields as part of the field, never as terminators. Doc comment updated to spell out the contract; new tests cover scattered blank lines, multiline quoted fields, and escaped quotes.

File: src/ui/src/utils/csvParse.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    } else if (ch === "\r") {
+      // \r\n handled by the \n branch on the next iteration.
+      if (input[i + 1] !== "\n") {
+        count++;
+        sawContent = false;
+      }
+    } else {
+      sawContent = true;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/utils/csvParse.ts]: Addressed in a52bc70 — adopted the spirit of the suggested implementation: `countCsvRows` (renamed from `countCsvLines` to make the contract explicit) now runs the same quote-aware state machine as `parseCsv`, so newlines inside `"..."` fields are not row terminators and the suggested escape-quote handling is in place. New tests cover the multiline-quoted-field case to lock the behavior in.

File: src/ui/src/components/chat/CsvAttachmentCard.tsx [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+                    ))}
+                  </tr>
+                ))}
+              </tbody>
+            </table>
+          </div>
+          {parsed.totalLines > parsed.rows.length && (
+            <div className={styles.csvTruncated}>
+              + {parsed.totalLines - parsed.rows.length} more rows
+            </div>
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/components/chat/CsvAttachmentCard.tsx:82]: Addressed in a52bc70 — `CsvAttachmentCard` now uses the new quote-aware `countCsvRows` (alongside `parseCsv`), so the row count and the parsed preview share the same model. Multiline quoted fields and scattered blank lines no longer skew the "+ N more rows" footer.

File: src/agent_mcp/tools/send_to_user.rs [hunk 5]
```diff
@@ -77,22 +86,43 @@ pub fn policy(media_type: &str, size_bytes: u64, filename: &str) -> Result<(), S
...
+            Some(m) => m,
+            None => {
+                return Err(format!(
+                    "media type {media_type:?} is not supported inline. Supported: \
+                     images (png/jpeg/gif/webp/svg), application/pdf, text/plain, \
+                     text/csv, text/markdown, application/json. The file at \
+                     {filename:?} is on disk — for unsupported types, just tell \
+                     the user that path so they can retrieve it manually instead \
+                     of calling this tool."
+                ));
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/agent_mcp/tools/send_to_user.rs:106]: Addressed in a52bc70 — both rejection strings were reworded to match the actual contract: `filename` is just the basename (the bridge strips the path), so the messages now say "instead tell the user the absolute path you wrote the file to" instead of implying that filename itself is the path. The model already has the absolute `file_path` it received as the tool argument, so this points it at what it has rather than at a basename. Test `rejection_text_is_actionable` updated to assert the new wording.

File: src/agent_mcp/tools/send_to_user.rs [hunk 6]
```diff
@@ -77,22 +86,43 @@ pub fn policy(media_type: &str, size_bytes: u64, filename: &str) -> Result<(), S
...
 
+/// Per-text-type cap, or `None` if the type isn't a recognized text/data type.
+fn max_text_bytes_for(media_type: &str) -> Option<u64> {
+    match media_type {
+        "text/plain" => Some(MAX_TEXT_BYTES),
+        "text/csv" => Some(MAX_CSV_BYTES),
+        "text/markdown" => Some(MAX_MARKDOWN_BYTES),
+        "application/json" => Some(MAX_JSON_BYTES),
+        _ => None,
+    }
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/agent_mcp/tools/send_to_user.rs]: Addressed in a52bc70 — consolidated into `TEXT_TYPE_RULES: &[(&str, u64)]` as the single source of truth. `policy()` looks the cap up in the table via `max_text_bytes_for`, and `server.rs` derives the JSON-Schema enum from `allowed_text_types()` (which iterates the same table). Added a `text_type_rules_round_trip` test to lock in the invariant. Adding/removing a type now requires only one edit.

File: src/ui/src/components/chat/MessageAttachment.tsx [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+/** Whether the given media type is rendered by `<MessageAttachment>`. The
+ *  caller should defer non-matching types to its own image / PDF branches. */
+export function isTextDataMediaType(mediaType: string): boolean {
+  return (
+    mediaType === "text/csv" ||
+    mediaType === "text/markdown" ||
+    mediaType === "application/json" ||
+    mediaType === "text/plain"
+  );
+}
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/components/chat/MessageAttachment.tsx:101]: Addressed in a52bc70 — went the "tighten the comments" direction since an implicit fallback would defeat the deliberate type-aware preview design (each card has different layout/parsing requirements). The doc comments on `MessageAttachment`, `isTextDataMediaType`, and `SUPPORTED_TEXT_TYPES` now spell out that there is no implicit fallback — adding a new text/data type requires touching the `switch`, `isTextDataMediaType`, and the inbound MIME allow-list together.

File: src/ui/src/utils/filePathLinks.ts [hunk 8]
```diff
@@ -0,0 +1,97 @@
+/**
+ * Detect absolute file paths inside a plain-text string so a markdown
+ * post-processor can wrap them in clickable links. The link target uses
+ * a custom `claudette-path:` URI so the markdown `<a>` override knows to
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/utils/filePathLinks.ts]: Addressed in e94bfa0 — updated the module header comment to match the actual `claudettepath:` scheme name.

File: src/ui/src/utils/filePathLinks.ts [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+export function encodeFilePathHref(path: string): string {
+  // encodeURI keeps slashes/backslashes legible in the DOM (helpful for
+  // debugging); only space and the few characters that genuinely break
+  // a URI need escaping.
+  return `${FILE_PATH_SCHEME}${encodeURI(path)}`;
+}
+
+export function decodeFilePathHref(href: string): string | null {
+  if (!href.startsWith(FILE_PATH_SCHEME)) return null;
+  return decodeURI(href.slice(FILE_PATH_SCHEME.length));
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/utils/filePathLinks.ts]: Addressed in e94bfa0 — `decodeFilePathHref` now wraps `decodeURI` in `try/catch` and returns the raw tail on `URIError` so a malformed `claudettepath:` link from assistant markdown cannot crash the markdown render or click handler. Added a regression test in `filePathLinks.test.ts`.

File: src/ui/src/utils/markdown.ts [hunk 10]
```diff
@@ -82,6 +85,10 @@ export function preprocessContent(text: string): string {
 }
 
 // Sanitization schema: allow standard markdown HTML + our callout elements.
+// Also extend `protocols.href` so the file-path autolinker's
+// `claudette-path:` scheme survives sanitization — without this, every
+// rehype-injected `<a href="claudette-path:…">` would be stripped to a bare
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/utils/markdown.ts]: Addressed in e94bfa0 — comments in `markdown.ts` now reference the actual `claudettepath:` scheme.

File: src/ui/src/utils/markdown.ts [hunk 11]
```diff
@@ -93,23 +100,79 @@ export const SANITIZE_SCHEMA = {
     div: [...(defaultSchema.attributes?.div ?? []), "className"],
     span: [...(defaultSchema.attributes?.span ?? []), "className"],
     hr: [...(defaultSchema.attributes?.hr ?? []), "className"],
+    // Default schema only allows `className` on <a> when the value is the
+    // hard-coded "data-footnote-backref" sentinel (footnote-specific). The
+    // file-path autolinker emits `cc-file-path-link`, which would otherwise
+    // get stripped — strip the value restriction by replacing the existing
+    // ["className", ...] entry with a bare "className" allow-all, then add
+    // back the original entries.
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/utils/markdown.ts]: Addressed in e94bfa0 — reworded the comment so it accurately describes what the code does: drop the existing value-restricted `["className", "data-footnote-backref"]` entry, then append a bare `"className"` so any value passes. Other a-attributes flow through via the spread.

File: src/ui/src/utils/rehypeFilePathLinks.ts [hunk 12]
```diff
@@ -0,0 +1,68 @@
+/**
+ * Rehype plugin that scans text nodes for absolute file paths and rewrites
+ * each match as an `<a href="claudette-path:…">` element. The MARKDOWN_COMPONENTS.a
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/utils/rehypeFilePathLinks.ts]: Addressed in e94bfa0 — module header now references `claudettepath:` (no hyphen) to match the constant.

File: src-tauri/src/agent_mcp_sink.rs [hunk 13]
```diff
@@ -28,10 +28,17 @@ use tauri::{AppHandle, Emitter, Manager};
 use crate::state::AppState;
 
 /// Tauri-side implementation of [`Sink`] — one per persistent agent session.
+///
+/// `chat_session_id` is the key used to look up the agent's `AgentSessionState`
+/// in `AppState.agents` (and therefore the `last_user_msg_id` anchor).
+/// `workspace_id` is what the frontend listens for on the
+/// `agent-attachment-created` event so it can route the new attachment to the
+/// right chat surface. Both are needed; they are NOT the same id.
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src-tauri/src/agent_mcp_sink.rs]: Addressed in e94bfa0 — reworded the doc-comment on `ChatBridgeSink` to describe the current routing: `chat_session_id` is now the key the frontend listener uses to merge attachments into `chatAttachments[sessionId]`. `workspace_id` is still emitted on the event but is no longer the routing key.

File: .github/workflows/ci.yml [hunk 14]
```diff
@@ -69,70 +69,10 @@ jobs:
...
-      - name: cargo test (library crates)
-        shell: bash
-        run: cargo test -p claudette -p claudette-server --target ${{ matrix.rust-target }}
+  # Note: Windows is not built in this CI pipeline. Windows artifacts
+  # are produced by the nightly (`.github/workflows/nightly.yml`) and
+  # release (`.github/workflows/release-please.yml`) workflows, which
+  # cover both `x86_64-pc-windows-msvc` and `aarch64-pc-windows-msvc`.
+  # Skipping Windows here keeps PR turnaround fast — Windows runners
+  # take ~10 min — without losing build-verification coverage where
+  # it matters (artifact production).
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply .github/workflows/ci.yml:78]: Not applicable — removing the PR-time `windows-check` job was an explicit policy call (see commit d843e92). Windows artifacts and verification still run on the nightly (`.github/workflows/nightly.yml`) and release (`.github/workflows/release-please.yml`) workflows, which build the actual `x86_64-pc-windows-msvc` and `aarch64-pc-windows-msvc` binaries. The PR-time job would only run a thin `cargo check` against `claudette-tauri` on Windows runners (~10 min per PR) for marginal additional signal beyond the existing fmt/clippy/test jobs that catch most regressions cross-platform. Keeping PR turnaround fast was the explicit goal here. (See the comment now retained in `ci.yml` at the spot where the job used to live.)

File: src/ui/src/components/chat/useAttachmentText.ts [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+    if (text !== null) return;
+    let cancelled = false;
+    (async () => {
+      try {
+        let b64 = data_base64 ?? null;
+        if (!b64 && attachmentId) {
+          const { loadAttachmentData } = await import("../../services/tauri");
+          b64 = await loadAttachmentData(attachmentId);
+        }
+        if (cancelled || !b64) return;
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/components/chat/useAttachmentText.ts]: Addressed in e94bfa0 — when neither inline `data_base64` nor a fetchable `attachmentId` is available the hook now sets an explicit error instead of returning silently, so the card surfaces the standard error state rather than getting stuck on "Loading…".

File: src-tauri/src/commands/chat.rs [hunk 16]
```diff
@@ -2489,49 +2539,86 @@ pub async fn read_file_as_base64(path: String) -> Result<AttachmentResponse, Str
...
+                text_content: None,
+                width: None,
+                height: None,
+                size_bytes,
+                origin: claudette::model::AttachmentOrigin::User,
+                tool_use_id: None,
+            })
+        }
+        Some((Kind::Csv | Kind::Markdown | Kind::Json, media_type)) => {
+            // Text-shaped — must be valid UTF-8.
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src-tauri/src/commands/chat.rs]: Addressed in e94bfa0 — `read_file_as_base64` now applies the same NUL-byte heuristic to the known text-shaped kinds (CSV / Markdown / JSON) that the unknown-extension branch already used. UTF-8 itself permits `U+0000`, so a binary file with a misleading suffix could pass `from_utf8` while clearly being binary; the prefix-window check rejects those uniformly now.

File: src/ui/src/utils/filePathLinks.ts [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+/**
+ * Sentence-final characters that tend to follow a path in prose but are
+ * almost never part of a real filename. Stripped from the tail of every
+ * match. Backticks are included so a path written like `` `…` `` doesn't
+ * absorb the closing backtick when it appears inside inline-code-adjacent
+ * prose. */
+const TRAILING_PUNCT_REGEX = /[.,;:!?)\]'"`]+$/;
+
+/** Real paths have at least one separator inside the body, not just the
+ *  prefix. `/a` is too short to be a useful match. */
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/utils/filePathLinks.ts]: Addressed in a395188 — reworded the `MIN_PATH_LENGTH` comment to match what the check actually does (a minimum-length heuristic, not an inner-separator requirement). The previous wording was misleading; tightening to "must contain another separator after the prefix" would have rejected legitimate paths like `/etc` and `~/foo`, so the heuristic itself stays put.

File: src/ui/src/utils/markdown.ts [hunk 18]
```diff
@@ -221,16 +285,32 @@ export function HighlightedCode({
...
+        // already, but title= adds redundancy on long paths that get
+        // visually truncated by surrounding wrap/ellipsis rules.
+        title: filePath ?? (props as { title?: string }).title,
         onClick: (e: React.MouseEvent<HTMLAnchorElement>) => {
+          if (filePath) {
+            e.preventDefault();
+            void invoke("open_in_editor", { path: filePath }).catch(
+              (err) =>
+                console.error("Failed to open path:", filePath, err),
+            );
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/utils/markdown.ts:311]: Addressed in a395188 — `open_in_editor` no longer detaches via `tauri::async_runtime::spawn`. It now awaits `opener::open` and returns the error to the frontend, so the markdown click handler's `.catch()` actually fires when the OS handler cannot open the path. The OS helpers (`open` / `xdg-open` / `cmd start`) all return as soon as they hand off to the registrar, so this stays non-blocking from the user's perspective.

File: src/ui/src/utils/filePathLinks.ts [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+ *    POSIX or NTFS act as natural terminators
+ *  - the regex stays compact enough to read
+ *
+ * Negative lookbehind on `[\w:.\\\/-]` blocks matches that start in the
+ * middle of an existing path or URL (e.g. `https://x.com/y` — the `/y`
+ * is preceded by `m`, fails). This is the cheapest way to avoid mangling
+ * URLs without a separate URL parser.
+ */
+const PATH_REGEX =
+  /(?<![\w:.\\/\-])(?:[A-Za-z]:[\\/][^\s<>"|*?]+|\\\\[^\s<>"|*?\\/]+\\[^\s<>"|*?\\/]+(?:\\[^\s<>"|*?\\/]+)*|~?\/[^\s<>"|*?]+)/g;
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/utils/filePathLinks.ts]: Addressed in a395188 — `open_in_editor` now expands a leading `~/` (or bare `~`) via `dirs::home_dir()` before passing the path to `opener::open`. Added four unit tests covering `~/foo`, `~`, absolute paths (POSIX + Windows-drive), and the `~user/...` case (which is intentionally not expanded — too rare to be worth platform-specific handling, and the OS opener will give a sensible error).

File: src/ui/src/utils/filePathLinks.ts [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/utils/filePathLinks.ts:60]: Addressed in 8ffa96e — confirmed `tauri.conf.json` sets `minimumSystemVersion: "11.0"` and WebKit lookbehind support only landed in 16.4 (≈ macOS 13.3), so this was a real top-level-eval crash for Big Sur users. Removed the `(?<![\w:.\\/\-])` lookbehind from `PATH_REGEX`; the boundary check now runs as an explicit JS-side guard against the preceding character (`FORBIDDEN_PREV_CHAR_REGEX`) inside `detectFilePaths`. Same semantics, but the module now evaluates cleanly on JSC < 16.4. Added a smoke test that asserts the function source contains no `(?<…)` syntax to regression-protect future edits.

File: src-tauri/src/commands/shell.rs [hunk 21]
```diff
@@ -36,13 +36,33 @@ pub fn detect_user_shell() -> (String, ShellType) {
...
+    // `opener::open` shells out to the OS handler, which doesn't perform
+    // shell-style tilde expansion — without this, the markdown autolinker
+    // emitting `~/Downloads/foo.csv` would silently fail to open.
+    let expanded = expand_home_tilde(&path);
+    // Run synchronously so errors propagate back to the frontend instead
+    // of getting silently dropped on stderr inside a spawned task. The
+    // helper is a thin wrapper around `open` / `xdg-open` / `cmd start`,
+    // all of which return as soon as the OS hands off — no blocking.
+    opener::open(&expanded).map_err(|e| format!("Failed to open {expanded:?}: {e}"))
+}
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src-tauri/src/commands/shell.rs:55]: Addressed in 8ffa96e — added `is_acceptable_open_target` which rejects (a) anything matching the `<scheme>://…` shape (catching `https://`, `file://`, `javascript://` and friends) and (b) relative or empty paths. Only POSIX absolute (`/foo`), Windows drive (`C:\` / `C:/`), and UNC (`\\server\…`) prefixes pass through. Validation runs after `expand_home_tilde` so a tilde-expanded `~/foo` still works. Six unit tests cover the accept and reject branches.
